x/wasm/internal/keeper/querier_test.go

```diff
@@ -0,0 +1,131 @@
+
+	wasmCode, err := ioutil.ReadFile("./testdata/contract.wasm")
+	require.NoError(t, err)
+
+	contractID, err := keeper.Create(ctx, creator, wasmCode)
+	require.NoError(t, err)
+
+	_, _, bob := keyPubAddr()
+	initMsg := InitMsg{
+		Verifier:    anyAddr.String(),
+		Beneficiary: bob.String(),
+	}
+	initMsgBz, err := json.Marshal(initMsg)
+	require.NoError(t, err)
+
+	addr, err := keeper.Instantiate(ctx, creator, contractID, initMsgBz, deposit)
+	require.NoError(t, err)
+	require.Equal(t, "cosmos18vd8fpwxzck93qlwghaj6arh4p7c5n89uzcee5", addr.String())
+
+	contractModel := []types.Model{
+		{Key: "foo", Value: "bar"},
+		{Key: string([]byte{0x0, 0x1}), Value: string([]byte{0x2, 0x3})},
+	}
+	keeper.setContractState(ctx, addr, contractModel)
+
+	q := NewQuerier(keeper)
+	specs := map[string]struct {
+		srcPath          []string
+		srcReq           abci.RequestQuery
+		expModelLen      int
+		expModelContains []model
+		expErr           sdk.Error
+	}{
+		"query all": {
+			srcPath:     []string{QueryGetContractState, addr.String(), QueryMethodContractStateAll},
+			expModelLen: 3,
+			expModelContains: []model{
+				{Key: "foo", Value: "bar"},
+				{Key: string([]byte{0x0, 0x1}), Value: string([]byte{0x2, 0x3})},
+			},
+		},
+		"query raw key": {
+			srcPath:          []string{QueryGetContractState, addr.String(), QueryMethodContractStateRaw},
+			srcReq:           abci.RequestQuery{Data: []byte("foo")},
+			expModelLen:      1,
+			expModelContains: []model{{Key: "foo", Value: "bar"}},
+		},
+		"query raw binary key": {
+			srcPath:          []string{QueryGetContractState, addr.String(), QueryMethodContractStateRaw},
+			srcReq:           abci.RequestQuery{Data: []byte{0x0, 0x1}},
+			expModelLen:      1,
+			expModelContains: []model{{Key: string([]byte{0x0, 0x1}), Value: string([]byte{0x2, 0x3})}},
+		},
+		"query smart": {
+			srcPath:     []string{QueryGetContractState, addr.String(), QueryMethodContractStateSmart},
+			srcReq:      abci.RequestQuery{Data: []byte(`{"raw":{"key":"config"}}`)},
+			expModelLen: 1,
+			//expModelContains: []model{}, // stopping here as contract internals are not stable
+		},
+		"query unknown raw key": {
+			srcPath:          []string{QueryGetContractState, addr.String(), QueryMethodContractStateRaw},
+			srcReq:           abci.RequestQuery{Data: []byte("unknown")},
+			expModelLen:      1,
+			expModelContains: []model{{Key: "unknown", Value: ""}},
+		},
+		"query empty raw key": {
+			srcPath:     []string{QueryGetContractState, addr.String(), QueryMethodContractStateRaw},
+			expModelLen: 0,
+		},
+		"query raw with unknown address": {
+			srcPath:     []string{QueryGetContractState, anyAddr.String(), QueryMethodContractStateRaw},
+			expModelLen: 0,
+		},
+		"query all with unknown address": {
+			srcPath:     []string{QueryGetContractState, anyAddr.String(), QueryMethodContractStateAll},
+			expModelLen: 0,
+		},
+		"query smart with unknown address": {
```

Comment:
I would assume all the "query with unknown address" would return the same `types.ErrNotFound("contract")` message

Reply:
Agree `ErrNotFound ` would be nicer but the `prefix.Store` concatenates the prefix and key for the parent store so that it results in just another unknown key.
I would need to do a prefix scan on the parent to detect the issue. The previous implementation for `all` had the same "problem"